Source/WebKit/UIProcess/Extensions/WebExtensionDeclarativeNetRequestSQLiteStore.cpp

```diff
@@ -0,0 +1,354 @@
+
+        rows = statement->fetch();
+    } else
+        rows = SQLiteDatabaseFetch(*(database()), makeString("SELECT * FROM "_s, m_tableName));
+
+    return getKeysAndValuesFromRowIterator(*rows);
+}
+
+Ref<JSON::Array> WebExtensionDeclarativeNetRequestSQLiteStore::getKeysAndValuesFromRowIterator(Ref<WebExtensionSQLiteRowEnumerator> rows)
+{
+    Ref results = JSON::Array::create();
+
+    RefPtr row = rows->next();
+    while (row != nullptr) {
+        auto rule = row->getData(1);
+
+        if (!rule)
+            continue;
+
+        auto ruleObject = JSON::Value::optionalParseJSON(StringView(rule->span()));
+
+        if (ruleObject)
+            results->pushValue(ruleObject->get());
+
+        row = rows->next();
+    }
+
+    return results;
+}
+
+String WebExtensionDeclarativeNetRequestSQLiteStore::insertRule(const JSON::Object& rule, Ref<WebExtensionSQLiteDatabase> database)
+{
+    assertIsCurrent(queue());
+
+    auto ruleData = rule.toJSONString();
+    auto ruleID = rule.getInteger("id"_s);
+    ASSERT(ruleID);
+
+    DatabaseResult result = SQLiteDatabaseExecute(database, makeString("INSERT INTO "_s, m_tableName, " (id, rule) VALUES (?, ?)"_s), *ruleID, ruleData);
+    if (result != SQLITE_DONE) {
+        RELEASE_LOG_ERROR(Extensions, "Failed to insert dynamic declarative net request rule for extension %s", uniqueIdentifier().utf8().data());
+        return makeString("Failed to add "_s, m_storageType, " rule."_s);
+    }
+
+    return nullString();
+}
+
+// MARK: Database Schema
+
+SchemaVersion WebExtensionDeclarativeNetRequestSQLiteStore::currentDatabaseSchemaVersion()
+{
+    return currentDeclarativeNetRequestSchemaVersion;
+}
+
+URL WebExtensionDeclarativeNetRequestSQLiteStore::databaseURL()
+{
+    if (useInMemoryDatabase())
+        return WebExtensionSQLiteDatabase::inMemoryDatabaseURL();
+
+    ASSERT(m_storageType == WebExtensionDeclarativeNetRequestStorageType::Dynamic);
+
+    String databaseName = "DeclarativeNetRequestRules.db"_s;
+
+    ASSERT(!directory().isEmpty());
+
+    return URL(URL { makeString(directory().string(), "/"_s) }, databaseName);
+}
+
+DatabaseResult WebExtensionDeclarativeNetRequestSQLiteStore::createFreshDatabaseSchema()
+{
+    assertIsCurrent(queue());
+    ASSERT(database());
+
+    DatabaseResult result = SQLiteDatabaseExecute(*database(), makeString("CREATE TABLE "_s, m_tableName, " (id INTEGER PRIMARY KEY NOT NULL, rule BLOB NOT NULL)"_s));
+    if (result != SQLITE_DONE)
+        RELEASE_LOG_ERROR(Extensions, "Failed to create %s database for extension %s: %s (%d)", m_tableName.utf8().data(), uniqueIdentifier().utf8().data(), lastErrorMessage().data(), result);
+    return result;
+}
+
+SchemaVersion WebExtensionDeclarativeNetRequestSQLiteStore::migrateToCurrentSchemaVersionIfNeeded()
+{
+    assertIsCurrent(queue());
+
+    auto currentDatabaseSchemaVersion = databaseSchemaVersion();
+    if (currentDatabaseSchemaVersion == 1 && currentDeclarativeNetRequestSchemaVersion == 2) {
```

Comment:
I think we should drop the `currentDeclarativeNetRequestSchemaVersion == 2`. What happens when the current versions is 3 and someone is on version 1?

```suggestion
    if (currentDatabaseSchemaVersion == 1) {
```